docs/core/forms.rst

```diff
@@ -8,278 +8,404 @@ Forms
+    from rasa_sdk.events import SlotSet
+
+
+    class ValidateSlots(Action):
+        def name(self) -> Text:
+            return "action_validate_your_form"
+
+        def run(
+            self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
+        ) -> List[EventType]:
+            extracted_slots: Dict[Text, Any] = tracker.get_extracted_slots()
+
+            validation_events = []
+
+            for slot_name, slot_value in extracted_slots:
+                # Check if slot is valid.
+                if is_valid(slot_value):
+                    validation_events.append(SlotSet(slot_name, slot_value))
+                else:
+                    # Return a `SlotSet` event with value `None` to indicate that this
+                    # slot still needs to be filled.
+                    validation_events.append(SlotSet(slot_name, None))
+
+            return validation_events
+
+        def is_valid(slot_value: Any) -> bool:
+            # Implementation of the validate function.
+
+
+.. _forms-custom-slot-mappings:
+
+Custom Slot Mappings
+~~~~~~~~~~~~~~~~~~~~
+
+If none of the predefined :ref:`forms-slot-mappings` fit your use case, you can use the
+:ref:`custom action<custom-actions>` ``action_validate_{form_name}`` to write your own
+extraction code. Rasa Open Source will trigger this function when the form is run.
+
+Make sure your custom action returns ``SlotSet`` events for every extracted value.
+The following example shows the implementation of a custom slot mapping which sets
+a slot based on the length of the last user message.
+
+.. code-block:: python
+
+    from typing import Dict, Text, List
+
+    from rasa_sdk import Tracker
+    from rasa_sdk.events import EventType
+    from rasa_sdk.executor import CollectingDispatcher
+    from rasa_sdk import Action
+    from rasa_sdk.events import SlotSet
+
+
+    class ValidateSlots(Action):
+        def name(self) -> Text:
+            return "action_validate_your_form"
+
+        def run(
+            self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
+        ) -> List[EventType]:
+            text_of_last_user_message = tracker.latest_message.get("text")
+
+            return [SlotSet("user_message_length", len(text_of_last_user_message))]
+
+
+Requesting Extra Slots
+~~~~~~~~~~~~~~~~~~~~~~
+
+If you have frequent changes to the required slots and don't want to retrain your
+assistant when something changes, you can also use a
+:ref:`custom action<custom-actions>` ``action_validate_{form_name}`` to define
+which slots should be requested. Rasa Open Source will run your custom action whenever
+the form is executed. Set the slot ``requested_slot`` to the name of the slot which
+should be extracted next. If all desired slots are filled, set ``requested_slot``
+to ``None``.
+
+The following example shows the implementation of a custom action which requests
+the three slots ``last_name``, ``first_name``, and ``city``.
+
```

Comment:
this example might be confusing, because it repeats the default behavior. @akelad what do you think?

Reply:
What do you mean with `because it repeats the default behavior` ? It's an example for requesting slots from a custom action which should only be done if having the slot mappings in the domain doesn't cut it for you.

Reply:
yes, but this particular example can be covered by putting these slots in the domain. So users who read it might think they need to do it anyhow

Reply:
It's in the `advanced usage` section and then I say

> If you have frequent changes to the required slots and don't want to retrain your
assistant when something in your form changes, you can also use a
:ref:`custom action<custom-actions>` ``validate_{form_name}`` to define
which slots should be requested.

I think it should be pretty clear

Reply:
but people might first look into example, so I suggest to create example that illustrates what is written there